File: bench/sort.c
/* Sorting benchmark: fixed pseudo-shuffle, qsort ascending, order-sensitive
   checksum. Same input and sorted order as sort.zeph/.rs => same checksum. */
#include <stdio.h>
#include <stdlib.h>

static int cmp(const void *a, const void *b) {
    long long x = *(const long long *)a, y = *(const long long *)b;
    return (x > y) - (x < y);
}

int main(int argc, char **argv) {
    long N = argc > 1 ? atol(argv[1]) : 3000000;
    long long *a = malloc((size_t)N * sizeof(long long));
    for (long i = 0; i < N; i++)
        a[i] = ((long long)i * 2654435761LL + 12345) % 1000003;
    qsort(a, (size_t)N, sizeof(long long), cmp);
    long long chk = 0;
    for (long i = 0; i < N; i++)
        chk = (chk * 31 + a[i]) % 1000000007;
    printf("%lld\n", chk);
    return 0;
}

File: bench/sort.rs
// Sorting benchmark: fixed pseudo-shuffle, sort_unstable ascending,
// order-sensitive checksum. Matches sort.zeph/.c.
use std::env;

fn main() {
    let ar: Vec<String> = env::args().collect();
    let n: i64 = if ar.len() > 1 { ar[1].parse().unwrap() } else { 3000000 };
    let mut a: Vec<i64> = (0..n).map(|i| (i * 2654435761 + 12345) % 1000003).collect();
    a.sort_unstable();
    let mut chk: i64 = 0;
    for i in 0..n as usize {
        chk = (chk * 31 + a[i]) % 1000000007;
    }
    println!("{}", chk);
}
